### PR TITLE
[NCL-7201] Implement SCM Repository Create and Edit page

### PR DESCRIPTION
**Pull request by DnsZhou**

### CODE REVIEW

**Review by matedo1 — DISMISSED** [review]

File: src/utils/formValidationHelpers.ts [hunk 1]
```diff
@@ -3,7 +3,11 @@
 // url regex taken from:
 // https://uibakery.io/regex-library/url
 // eslint-disable-next-line
-const urlRegex = /^https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&\/=]*)$/;
+const urlRegex = /^https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_+.~#?&/=]*)$/;
```
[the remark this answers is not in the document]
Comment (DnsZhou) [reply src/utils/formValidationHelpers.ts:6]: Verified that the original regex was fine. reverted this line

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+      .run({
+        serviceData: {
+          data: {
+            scmUrl: data.scmUrl.value,
+            // preBuildSyncEnabled: data.preBuildSyncEnabled.value,
+          },
+        },
+      })
+      .then((response: any) => {
+        // @Todo: Verify the create result from the WS Message after WS was implemented.
```
[the remark this answers is not in the document]
Comment (DnsZhou) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: NCL-7935 created

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+  useEffect(() => {
+    if (isEditPage) {
+      if (urlPathParams.scmRepositoryId) {
+        serviceContainerEditPageGetRunner({ serviceData: { id: urlPathParams.scmRepositoryId } }).then((response: any) => {
+          const scmRepository: SCMRepository = response.data;
+
+          setScmRepository(scmRepository);
+          setId(scmRepository.id);
+          reinitialize({
+            internalUrl: scmRepository.internalUrl,
```
[the remark this answers is not in the document]
Comment (DnsZhou) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: removed

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+interface IScmRepositoryCreateEditPageProps {
+  isEditPage?: boolean;
+}
+
+const createFormConfig = {
+  scmUrl: {
+    isRequired: true,
+    validators: [{ validator: validateScmUrl, errorMessage: 'Invalid URL format.' }],
+  },
+  externalUrl: {
```
[the remark this answers is not in the document]
Comment (DnsZhou) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: removed

File: src/utils/formValidationHelpers.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/utils/formValidationHelpers.ts]: done

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+  },
+  // preBuildSyncEnabled: {},
+};
+
+export const ScmRepositoryCreateEditPage = ({ isEditPage = false }: IScmRepositoryCreateEditPageProps) => {
+  const flexDirection: FlexProps['direction'] = { default: 'column' };
+
+  const [id, setId] = useState<string>();
+  const [scmRepository, setScmRepository] = useState<SCMRepository>();
+  const [isSubmitLoading, setIsSubmitLoading] = useState<boolean>(false);
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: done

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+  );
+
+  return (
+    <PageLayout
+      title={isEditPage ? 'Update SCM Repository' : 'Create SCM Repository'}
+      description={
+        isEditPage ? (
+          <>You can update current SCM Repository attributes below.</>
+        ) : (
+          <>You can create a SCM Repository with either a Internal URL(from Gerrit) or External URL(from any other SCM system).</>
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: done

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+                }}
+              />
+            </FormGroup>
+          )}
+          {isEditPage && (
+            <>
+              <FormGroup label="Internal SCM URL" fieldId="internalUrl">
+                <ScmRepositoryUrl internalScmRepository={scmRepository} />
+              </FormGroup>
+              <FormGroup label="External SCM URL" fieldId="externalUrl">
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: done

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: done

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+              onChange={(preBuildSyncEnabled) => {
+                onChange('preBuildSyncEnabled', preBuildSyncEnabled);
+              }}
+            />
+          </FormGroup> */}
+          <ActionGroup>
+            <Button
+              variant="primary"
+              isDisabled={isSubmitDisabled}
+              isLoading={isSubmitLoading}
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: done (removed)

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: done

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx]: done

File: src/AppRoutes.tsx [hunk 9]
```diff
@@ -116,6 +117,22 @@ export const AppRoutes = () => (
...
+          <ProtectedRoute title={PageTitles.scmRepositoryCreate}>
+            <ScmRepositoryCreateEditPage />
+          </ProtectedRoute>
+        }
+      />
+      <Route
+        path=":scmRepositoryId/edit"
+        element={
+          <ProtectedRoute title={PageTitles.scmRepositoryEdit}>
+            <ScmRepositoryCreateEditPage isEditPage={true} />
```
Comment (patrikk0123) [inline src/AppRoutes.tsx:132]: just `isEditPage` is enough

File: src/services/scmRepositoryApi.ts [hunk 10]
```diff
@@ -27,3 +28,28 @@ export const getScmRepositories = (requestConfig: AxiosRequestConfig = {}) => {
...
+ *
+ * @param data - object containing new SCM Repository data
+ * @param requestConfig - Axios based request config
+ */
+export const createScmRepository = ({ data }: { data: Omit<SCMRepository, 'id'> }, requestConfig: AxiosRequestConfig = {}) => {
+  return pncClient.getHttpClient().post<SCMRepository>('/scm-repositories/create-and-sync', data, requestConfig);
+};
+
+/**
+ * Patch a SCM Repository.
```
Comment (patrikk0123) [inline src/services/scmRepositoryApi.ts:43]: an SCM

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+
+  return (
+    <PageLayout
+      title={isEditPage ? 'Edit SCM Repository' : 'Create SCM Repository'}
+      description={
+        isEditPage ? (
+          <>You can edit current SCM Repository attributes below.</>
+        ) : (
+          <>
+            You can create a SCM Repository like <label>apache/maven.git</label> or <label>git/twitter4j.git</label>. It can be
```
Comment (patrikk0123) [inline src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx:245]: an SCM

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+
+const editFormConfig = {
+  externalUrl: {
+    validators: [{ validator: validateScmUrl, errorMessage: 'Invalid SCM URL format.' }],
+  },
+  // preBuildSyncEnabled: {},
+};
+
+export const ScmRepositoryCreateEditPage = ({ isEditPage = false }: IScmRepositoryCreateEditPageProps) => {
+  const flexDirection: FlexProps['direction'] = { default: 'column' };
```
Comment (patrikk0123) [inline src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx:54]: `flexDirectionColumn` and move this outside the component

File: src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+                  <TooltipWrapper tooltip="URL to the internal SCM repository, which is the main repository used for the builds." />
+                }
+              >
+                {scmRepository && <ScmRepositoryUrl internalScmRepository={scmRepository} />}
+              </FormGroup>
+              <FormGroup
+                label="External SCM URL"
+                fieldId="externalUrl"
+                labelIcon={<TooltipWrapper tooltip="URL to the upstream SCM repository." />}
+              >
```
Comment (patrikk0123) [inline src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx:187]: This FormGroup misses the `helperText` and `isRequired`

Comment (DnsZhou) [reply src/components/ScmRepositoryCreateEditPage/ScmRepositoryCreateEditPage.tsx:187]: HelperText added. External SCM URL can be empty when you editing the scm repo.

**Review by patrikk0123 — APPROVED** [review]

**Pull request merged**